common/achievement/evaluate.ts

```diff
@@ -0,0 +1,117 @@
+import { Achievement_event } from '../../graphql/generated';
+import { BucketConfig, BucketEvents, BucketEventsWithAggr, ConditionDataAggregations, EvaluationResult } from './types';
+import { prisma } from '../prisma';
+import { aggregators } from './aggregator';
+import swan from '@onlabsorg/swan-js';
+import { bucketCreatorDefs } from './bucket';
+
+export async function evaluateAchievement(condition: string, dataAggregation: ConditionDataAggregations, metrics: string[]): Promise<EvaluationResult> {
+    const achievementEvents = await prisma.achievement_event.findMany({ where: { metric: { in: metrics } } });
+
+    const eventsByMetric: Record<string, Achievement_event[]> = {}; // Store events per metric
+    for (const event of achievementEvents) {
+        if (!eventsByMetric[event.metric]) {
+            eventsByMetric[event.metric] = [];
+        }
+        eventsByMetric[event.metric].push(event);
+    }
+
+    const resultObject: Record<string, number | string | boolean> = {};
+
+    for (const key in dataAggregation) {
+        if (!dataAggregation[key]) {
+            return;
+        }
+        const dataAggregationObject = dataAggregation[key];
+        const metricName = dataAggregationObject.metric;
+
+        const bucketCreator = dataAggregationObject.createBuckets || 'default';
+        const bucketAggregator = dataAggregationObject.bucketAggregator || 'count';
+
+        const aggregator = dataAggregationObject.aggregator;
+
+        const eventsForMetric = eventsByMetric[metricName];
+        if (!eventsForMetric) {
+            return;
+        }
+        // we take the relation from the first event, that posesses one, in order to create buckets from it, if needed
+        const relation = eventsForMetric.find((event) => event.relation)?.relation;
+
+        const bucketCreatorFunction = bucketCreatorDefs[bucketCreator].function;
+        const bucketAggregatorFunction = aggregators[bucketAggregator].function;
+        const aggFunction = aggregators[aggregator].function;
+
+        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggFunction) {
+            return;
```

Comment:
I would either throw an error in this case or stay with the other return format where `conditionIsMet: false`. In any case I would suggest to log an error with all the information needed as this is a misconfiguration on our side.

Reply:
I logg an error message containing the values that caused the error and also where the error is originating from.

Reply:
I let the function continue anyways, the evaluation will turn out false anyhow.

Reply:
Why not returning? It would make sure that we don't run into weird side-effects. Furthermore, it doesn't make sense to spend more cpu time when we already know that we don't like the result.

Reply:
good point.